pennylane_lightning/core/src/simulators/lightning_tensor/tncuda/MPOTNCuda.hpp

```diff
@@ -0,0 +1,220 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/**
+ * @file MPOTNCuda.hpp
+ * Class for cuTensorNet-backed Matrix Product Operator.
+ */
+
+#pragma once
+
+#include <algorithm>
+#include <cuComplex.h>
+#include <cutensornet.h>
+#include <vector>
+
+#include "TensorCuda.hpp"
+#include "tncudaError.hpp"
+#include "tncuda_helpers.hpp"
+
+namespace {
+namespace cuUtil = Pennylane::LightningGPU::Util;
+}
+
+namespace Pennylane::LightningTensor::TNCuda {
+
+/**
+ * @brief Class representing an Matrix Product Operator (MPO) object for the MPS
+ backend.
+ * Any gate tensor can be represented as an MPO tensor network in the context of
+ MPS. The gate tensor must be decomposed with respect to its target wires. If
+ the target wires are not adjacent, identity tensors are inserted between the
+ MPO tensors.
+ 1. The MPO tensors' modes order in an open boundary condition are:
+   2              3              2
+   |              |              |
+   X--1--....--0--X--2--....--0--X
+   |              |              |
+   0              1              1
+
+ 2. The extents of the MPO tensors are [bondL, 2, bondR, 2]. The bondL of the
+ left side bound MPO tensor is 1 and the bondR of the right side bound MPO
+ tensor is 1.
+
+ Note that the gate tensor should be permuted to ascending order and decomposed
+ into MPO sites before passing to this class. Preprocess and postprocess with
+ SWAP operations are required to ensure MPOs target at adjacent wires and the
+ target wires are correct.
+
+ * @tparam PrecisionT Floating point type.
+ */
+template <class PrecisionT> class MPOTNCuda {
+  private:
+    using ComplexT = std::complex<PrecisionT>;
+    using CFP_t = decltype(cuUtil::getCudaType(PrecisionT{}));
+
+    cutensornetNetworkOperator_t MPOOperator_;
+    cuDoubleComplex coeff_ =
+        make_cuDoubleComplex(1.0, 0.0); // default coefficient
+    cutensornetBoundaryCondition_t boundaryCondition_{
+        CUTENSORNET_BOUNDARY_CONDITION_OPEN}; // open boundary condition
+    int64_t componentIdx_;
+
+    std::vector<std::size_t> bondDims_;
+
+    std::size_t numMPOSites_;
+    std::vector<std::size_t> modes_;
+    std::vector<int32_t> MPO_modes_int32_;
+
+    std::vector<std::vector<int64_t>> modesExtents_int64_;
+    // TODO: Explore if tensors_ can be stored in a separate memory manager
+    // class
+    std::vector<std::shared_ptr<TensorCuda<PrecisionT>>> tensors_;
+
+    /**
+     * @brief Get a vector of pointers to extents of each site.
+     *
+     * @return std::vector<int64_t const *> Note int64_t const* is
+     * required by cutensornet backend.
+     */
+    [[nodiscard]] auto getModeExtentsPtr_() -> std::vector<int64_t const *> {
+        std::vector<int64_t const *> modeExtentsPtr_int64(numMPOSites_);
+        for (std::size_t i = 0; i < numMPOSites_; i++) {
+            modeExtentsPtr_int64[i] = modesExtents_int64_[i].data();
+        }
+        return modeExtentsPtr_int64;
```

Comment:
```suggestion
        std::transform(modesExtents_int64_.begin(), modesExtents_int64_.end(), modeExtentsPtr_int64.begin(), [] (std::vector<int64_t> x){return x.data()})
        return modeExtentsPtr_int64;
```
I didn't test it, please test before committing. :slightly_smiling_face:

Reply:
Nice one. Instead of using `std::transform` (tried but failed), I applied iterator based for-loop, which can also avoid explicitly specifying the size of mode_extents_ptr vector and be easier for maintanence.